### ISSUE TITLE
DayOne: Editing existing entries with tags removes the tags when saving

### ISSUE THREAD

**Issue by flight16** [jrnl-org/jrnl#358]
1. Create a DayOne entry using DayOne.app and add a tag using the tag button at the top of the entry (don't include the tag in the body).
2. Click save and see that the entry is now tagged.
3. Close DayOne
4. Run jrnl dayone -n 1 --edit and edit the entry you just added, and save it.
5. Open DayOne and note that the latest entry is no longer tagged.

**Comment by flight16** [comment]
I think this might have been caused by fd46ffea23eb1cb98121bde3a3993a4fa56e1afb, but I haven't had time to investigate yet.

I think this shows there are two tests that are lacking:
0. A test to flex the --edit code path which launches an external editor and make sure data round-trips OK.
0. A test to verify the format of saved DayOne .xml to make sure it conforms to DayOne's expected format.

**Comment by MinchinWeb** [comment]
I've been able to confirm the issue.

I think it is somewhat related to #351.

It actually exists much earlier. It exists as of ea6060d and as of 3912693, which dates to last September. Which makes me wonder if it has ever worked...

---

The problem seems to be here -> [DayOneJournal.py](https://github.com/maebert/jrnl/blob/8a6563d958776cef47fe3f88564602464319f18f/jrnl/DayOneJournal.py#L132) which calls this function -> [Entry.py](https://github.com/maebert/jrnl/blob/8a6563d958776cef47fe3f88564602464319f18f/jrnl/Entry.py#L25)

---

Does 8f12d62 fix it?

**Comment by MinchinWeb** [comment]
waiting on merge of PR #350

**Comment by wren** [comment]
@MinchinWeb Sorry, but it looks like #350 has merge conflicts again.

Honestly, I'm having some trouble with what to do with Day One functionality. I'm hoping we can talk it through a bit, since it seems to be an important feature for you.

The way I see it, the Day One functionality in jrnl needs some work to get it, and its tests, up to snuff. And then it also needs someone that can seriously commit to maintaining it long term. When I first started fixing up jrnl, I have to admit that the Day One functionality seemed like a top candidate for cutting out (for a variety of reasons that I'm happy to get into, if you're interested).

That being said, I want to respect the existing community of users, and especially existing contributors. If there's a feature that you value this much, I want to make sure it gets the attention it deserves.

I want to apologize for not being able to review your PRs in a timely manner. The main problem I'm having is that I have no way to run and/or test Day One Classic. And without it, I can't verify that a test entered into our current test suite works as intended. I also can't work on, or even review PRs that address Day One functionality in any meaningful way. So, whoever ends up fixing and maintaining the Day One functionality in jrnl, I don't think that person can be me.

All of that is to say, do you think that person can be you (or anyone else reading this)? Can you fix this part of jrnl and keep it maintained? If you want to make a serious commitment to do this, can we can talk through some steps to get this done? If not, that's perfectly fine. Either way, just let me know, and we can go from there.

**Comment by MinchinWeb** [comment]
Hi @wren,

Let me preface this discussion by saying that there are two versions of Day One: the "Classic" version, and the current version, "DayOne 2". The Classic version allowed you to sync to Dropbox, which made the entries easily accessible to anywhere else you have your Dropbox synced to. v2 includes a sync feature, but seems to only be through an inaccessible, proprietary system. v2 was launched in 2016 and the Classic version was quickly thereafter sunsetted and removed from the App store. Dropbox changed their API about a year ago and so even if you somehow had access to the original app, it's unlikely the Dropbox sync would work today.

I'd be happy to do the work to keep support for DayOne Classic in *jrnl*. Let's leave the discussion for DayOne v2 for another day (as I don't use it). I don't actually have a way to run Day One Classic either, but I'm not sure it's really needed. Supporting a dead program like this has the added bonus that the file format should be entirely stable, warts and all. If you want more confidence that the test suite is actually testing what we're trying to do, we can work on expanding that.

What this issue is about is just being able to roundtrip correctly. That should be simple to build into the test suite (although I see it hasn't really been done in the case of #350). What I can say is that I've been running #350 locally for probably two years without issue.

Let me know what you think should be in place to feel more confident in our DayOne code.

**Comment by MinchinWeb** [comment]
The merge conflict for #350 has been resolved, all the tests pass, and should be good to merge!

**Comment by micahellison** [comment]
Hi @MinchinWeb, I read through #350 and I'm a little confused about this. What is your use case for integrating Journaley with jrnl?

**Comment by MinchinWeb** [comment]
Journaley is a GUI client for DayOne Classic on Windows. Because both Journaley and jrnl read the DayOne Classic data files, I can use both to interact with my "journal". I use Journaley as an editing client for my daily work journal, and I use jrnl to covert that journal into a personal-use website (via Pelican).

**Comment by wren** [comment]
@MinchinWeb Thanks for the explanation about everything. If we want to proceed, I'd like to take a few steps to make sure this functionality is fully supported, and also doesn't affect the rest of the project in any negative way.

I must confess that having a feature that nobody can test in any meaningful way makes me very weary. But I'm willing to proceed if you are willing to handle all of the Day One-related issues. This is the plan I would like to follow:

1. **Tests** - Get a PR with the full set of Day One tests reinstated and updated to fully pass on the current version of  jrnl. These should be separated into a Day One test file, and should, to the best of our ability, work on Mac and Windows. This PR should only have tests (no other code).
2. **PRs** - New PRs should be contained entirely in the `jrnl/DayOneJournal.py` file. If there needs to be any changes outside of that file, they should be filed as a separate PR, with an associated issue filed to justify the change. From then on, I will rubber-stamp any changes to `jrnl/DayOneJournal.py` that you approve without any further testing on my part.

What do you think?

**Comment by MinchinWeb** [comment]
@wren It sounds workable indeed; I'm happy to shepherd this along. If this doesn't work quite the way we haven imagined, we can always revisit our process. The only caveat that I would add is I intend to support only DayOne "Classic".

I'll start working on getting the pieces ready to cleanly merge. Is [2.5](https://github.com/jrnl-org/jrnl/tree/v2.5) the branch I should be targeting?

**Comment by wren** [comment]
@MinchinWeb Yup, v2.5 sounds good. You can submit PRs to master, though, and I'll update as needed to hit our milestones.

I agree about Classic. We're currently not in a place to support Day One 2. Would you mind updating the docs, to reflect that as part of this?

**Comment by MinchinWeb** [comment]
@wren #733 has been submitted to this end! Please let me know if there are any issues in getting it merged in.

**Comment by wren** [comment]
@MinchinWeb Is it possible for you to submit a PR first with just the tests for Day One, per the plan we agreed to above? It would make it a lot easier to review.

**Comment by MinchinWeb** [comment]
@wren : Created #742 with just DayOne tests! I didn't add any new tests, but just pulled in the ones that were previously removed.

**Comment by MinchinWeb** [comment]
bump

**Comment by MinchinWeb** [comment]
bump

**Comment by wren** [comment]
@MinchinWeb Thanks for the reminder. Are you still planning to submit the follow-up PR for this?

**Comment by MinchinWeb** [comment]
Yes, I intend to. The code is there in my previous omnibus PR's, I just need to figure out what code (from the omnibus PR) still needs to be merged in.

**Comment by wren** [comment]
Ah, okay. Just wanted to make sure you weren't waiting on me for anything. Let me know if you need anything.

**Comment by micahellison** [comment]
Looks like there's a PR in progress for this issue. Bumping to keep it from staling out.
